specreduce/extract.py

```diff
@@ -143,3 +145,108 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
+
+        variance : `~astropy.nddata.CCDData` or array-like, required
+            The associated variances for each pixel in the image.
+            Should have the same dimensions as `image`.
+
+        trace_object : `~specreduce.tracing.Trace`, required
+            The associated 1D trace object created for the 2D image.
+
+        disp_axis : int, optional
+            The index of the image's dispersion axis. [default: 1]
+
+        crossdisp_axis : int, optional
+            The index of the image's cross-dispersion axis. [default: 0]
+
+        bkgrd_prof : `~astropy.modeling.Model`
+            A model for the image's background flux.
+            [default: models.Polynomial1D(2)]
+
+        Returns
+        -------
+        spec_1d : `~specutils.Spectrum1D`
+            The final, Horne extracted 1D spectrum.
+        """
+        # co-add signal in each column
+        ncols = image.shape[crossdisp_axis]
+        xd_pixels = np.arange(ncols) # y plot dir / x spec dir
+        coadd = image.sum(axis=disp_axis) / ncols
+
+        # fit source profile, using Gaussian model as a template
+        # NOTE: could add argument for users to provide their own model
+        gauss_prof = models.Gaussian1D(amplitude=coadd.max(),
+                                       mean=coadd.argmax(), stddev=2)
+
+        # Fit extraction kernel to column with combined gaussian/bkgrd model
+        ext_prof = gauss_prof + bkgrd_prof
+        fitter = fitting.LevMarLSQFitter()
+        fit_ext_kernel = fitter(ext_prof, xd_pixels, coadd)
+
+        # generate 1D spectrum, column by column
+        extracted = []
+        for col_pix in range(image.shape[disp_axis]):
+            # set gaussian model's mean as column's corresponding trace value
+            fit_ext_kernel.mean_0 = trace_object.trace[col_pix]
+            # kernel_col.stddev_0 = self.fwhm_fit(x)
+            # NOTE: support for variable FWHMs forthcoming
+
+            # fit compound model to column
+            kernel_vals = fit_ext_kernel(xd_pixels)
+            # NOTE: support user-provided kernels? would require different handling
+            # NOTE: allow smoothing in x direction via boxcar/x axis collapse/etc?
+
+            # fetch matching columns from original and variance images
+            image_col = image[:, col_pix]
+            variance_col = variance[:, col_pix]
+
+            # calculate kernel normalization
+            g_x = np.ma.sum(kernel_vals**2 / variance_col)
+            if np.ma.is_masked(g_x):
+                continue
+
+            # sum by column weights
+            weighted_col = np.ma.divide(image_col * kernel_vals, variance_col)
+            result = np.ma.sum(weighted_col) / g_x
+
+            # multiply kernel normalization into the extracted signal
+            result *= (fit_ext_kernel.amplitude_0
+                       * fit_ext_kernel.stddev_0 * np.sqrt(2*np.pi))
+            extracted.append(result)
```

Comment:
if speed for these appends is a concern, this could be either vectorized or refactored into a list comprehension (by moving everything in the for-loop into an internal function, `_extract_col`, and calling `extracted = [_extract_col(col_pix) for col_pix in range(image.shape[disp_axis])]` although then the `continue` might cause issues.

Reply:
I got rid of the `continue` in the latest commit. Since the arrays are masked, any NaNs can just be carried on to the final, extracted 1D spectrum.